numpyro/distributions/directional.py

```diff
@@ -75,6 +132,246 @@ def variance(self):
+            - m * jnp.log(4 * jnp.prod(conc, axis=-1))
+        )
+        fs += log_I1(49, conc, terms=51).sum(-1)
+        mfs = fs.max()
+        norm_const = 2 * jnp.log(jnp.array(2 * pi)) + mfs + logsumexp(fs - mfs, 0)
+        return norm_const.reshape(self.phi_loc.shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        indv = self.phi_concentration * jnp.cos(
+            value[..., 0] - self.phi_loc
+        ) + self.psi_concentration * jnp.cos(value[..., 1] - self.psi_loc)
+        corr = (
+            self.correlation
+            * jnp.sin(value[..., 0] - self.phi_loc)
+            * jnp.sin(value[..., 1] - self.psi_loc)
+        )
+        return indv + corr - self.norm_const
+
+    def sample(self, key, sample_shape=()):
+        """
+        ** References: **
+            1. A New Unified Approach for the Simulation of aWide Class of Directional Distributions
+               John T. Kent, Asaad M. Ganeiber & Kanti V. Mardia (2018)
+        """
+        phi_key, psi_key = random.split(key)
+
+        corr = self.correlation
+        conc = jnp.stack((self.phi_concentration, self.psi_concentration))
+
+        eig = 0.5 * (conc[0] - corr ** 2 / conc[1])
+        eig = jnp.stack((jnp.zeros_like(eig), eig))
+        eigmin = jnp.where(eig[1] < 0, eig[1], jnp.zeros_like(eig[1], dtype=eig.dtype))
+        eig = eig - eigmin
+        b0 = self._bfind(eig)
+
+        total = _numel(sample_shape)
+        phi_den = log_I1(0, conc[1]).squeeze(0)
+        phi_shape = (total, 2, _numel(self.batch_shape))
+        phi_state = SineBivariateVonMises._phi_marginal(
+            phi_shape, phi_key, conc, corr, eig, b0, eigmin, phi_den
+        )
+
+        phi = lax.atan2(phi_state.phi[:, 1:], phi_state.phi[:, :1])
+
+        alpha = jnp.sqrt(conc[1] ** 2 + (corr * jnp.sin(phi)) ** 2)
+        beta = lax.atan(corr / conc[1] * jnp.sin(phi))
+
+        psi = VonMises(beta, alpha).sample(psi_key)
+
+        phi_psi = jnp.concatenate(
+            (
+                (phi + self.phi_loc + pi) % (2 * pi) - pi,
+                (psi + self.psi_loc + pi) % (2 * pi) - pi,
+            ),
+            axis=1,
+        )
+        phi_psi = jnp.transpose(phi_psi, (0, 2, 1))
+        return phi_psi.reshape(*sample_shape, *self.batch_shape, *self.event_shape)
+
+    @staticmethod
+    def _phi_marginal(shape, rng_key, conc, corr, eig, b0, eigmin, phi_den):
+        conc = jnp.broadcast_to(conc, shape)
+        eig = jnp.broadcast_to(eig, shape)
+        b0 = jnp.broadcast_to(b0, shape)
+        eigmin = jnp.broadcast_to(eigmin, shape)
+        phi_den = jnp.broadcast_to(phi_den, shape)
+
+        def update_fn(curr):
+            i, done, phi, key = curr
+            phi_key, key = random.split(key)
+            accept_key, acg_key, phi_key = random.split(phi_key, 3)
+
+            x = jnp.sqrt(1 + 2 * eig / b0) * random.normal(acg_key, shape)
+            x /= jnp.linalg.norm(x, axis=1)[
```

Comment:
I guess you wanted to use keepdims here?

Reply:
Yes, thanks!